ecl/hql/hqlgram2.cpp

```diff
@@ -9271,6 +9280,94 @@ bool HqlGram::checkAllowed(const attribute & errpos, const char *category, const
     return true;
 }
 
+bool HqlGram::checkDFSfields(IHqlExpression *dfsRecord, IHqlExpression *defaultRecord, const attribute& errpos)
+{
+    if (dfsRecord==defaultRecord)
+        return true;
+    bool ret = true;
+    IHqlSimpleScope *srcScope = dfsRecord->querySimpleScope();
+    ForEachChild(idx, defaultRecord)
+    {
+        IHqlExpression *subfield = defaultRecord->queryChild(idx);
+        switch (subfield->getOperator())
+        {
+        case no_record:
+            if (!checkDFSfields(dfsRecord, subfield, errpos))
+                ret = false;
+            break;
+        case no_ifblock:
+            if (!checkDFSfields(dfsRecord, subfield->queryChild(1), errpos))
+                ret = false;
+            break;
+        case no_field:
+            {
+                OwnedHqlExpr match = srcScope->lookupSymbol(subfield->queryId());
+                if (!match)
+                {
+                    reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s not found", str(subfield->queryId()));
+                    ret = false;
+                }
+                else if (subfield->getOperator() != match->getOperator())
+                {
+                    reportError(HQLERR_DFSlookupIncompatible, errpos, "Default record does not match resolved record - field %s incompatible", str(subfield->queryId()));
+                    ret = false;
+                }
+                else if (subfield->queryRecord() && match->queryRecord())
+                {
+                        ret = checkDFSfields(match->queryRecord(), match->queryRecord(), errpos) && ret;
```

Comment:
Does not catch error where a row is converted to a dataset/dictionary.  Should probably also check that match->queryType()->getTypeCode() is the same.